pkg/controllers/work/apply_controller.go

```diff
@@ -416,21 +442,117 @@ func (r *ApplyWorkReconciler) applyUnstructured(ctx context.Context, gvr schema.
+	if deployment.Status.ObservedGeneration == deployment.Generation && depCond != nil && depCond.Status == v1.ConditionTrue {
+		klog.V(2).InfoS("deployment is available", "deployment", klog.KObj(curObj))
+		return ManifestAvailableAction, nil
+	}
+	klog.V(2).InfoS("still need to wait for deployment to be available", "deployment", klog.KObj(curObj))
+	return ManifestNotAvailableYetAction, nil
 }
 
-// applyObject uses server side apply to apply the manifest.
-func (r *ApplyWorkReconciler) applyObject(ctx context.Context, gvr schema.GroupVersionResource,
+func trackStatefulSetAvailability(curObj *unstructured.Unstructured) (applyAction, error) {
+	var statefulSet appv1.StatefulSet
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(curObj.Object, &statefulSet); err != nil {
+		return ErrorApplyAction, controller.NewUnexpectedBehaviorError(err)
+	}
+	// a statefulSet is available if all the replicas are available and the currentReplicas is equal to the updatedReplicas
+	// which means there is no more update in progress.
+	if statefulSet.Status.ObservedGeneration == statefulSet.Generation &&
+		statefulSet.Status.AvailableReplicas == *statefulSet.Spec.Replicas &&
+		statefulSet.Status.CurrentReplicas == statefulSet.Status.UpdatedReplicas {
+		klog.V(2).InfoS("statefulSet is available", "statefulSet", klog.KObj(curObj))
+		return ManifestAvailableAction, nil
+	}
+	klog.V(2).InfoS("still need to wait for statefulSet to be available", "statefulSet", klog.KObj(curObj))
+	return ManifestNotAvailableYetAction, nil
+}
+
+func trackDaemonSetAvailability(curObj *unstructured.Unstructured) (applyAction, error) {
+	var daemonSet appv1.DaemonSet
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(curObj.Object, &daemonSet); err != nil {
+		return ErrorApplyAction, controller.NewUnexpectedBehaviorError(err)
+	}
+	// a daemonSet is available if all the desired replicas (equal to all node suit for this Daemonset)
+	// are available and the currentReplicas is equal to the updatedReplicas which means there is no more update in progress.
+	if daemonSet.Status.ObservedGeneration == daemonSet.Generation &&
+		daemonSet.Status.NumberAvailable == daemonSet.Status.DesiredNumberScheduled &&
+		daemonSet.Status.CurrentNumberScheduled == daemonSet.Status.UpdatedNumberScheduled {
+		klog.V(2).InfoS("daemonSet is available", "daemonSet", klog.KObj(curObj))
+		return ManifestAvailableAction, nil
+	}
+	klog.V(2).InfoS("still need to wait for daemonSet to be available", "daemonSet", klog.KObj(curObj))
+	return ManifestNotAvailableYetAction, nil
+}
+
+func trackJobAvailability(curObj *unstructured.Unstructured) (applyAction, error) {
+	var job batchv1.Job
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(curObj.Object, &job); err != nil {
+		return ErrorApplyAction, controller.NewUnexpectedBehaviorError(err)
+	}
+	if job.Status.Ready != nil {
```

Comment:
Hi Ryan! I am not 100% sure about the condition here: for the Job API, the pod might not stay in the ready stage; if it's no longer active (succeeded, failed), should we also consider it as available?

Reply:
For completed Jobs, TTL might also add some complications here -> though that might be less of a concern

Reply:
I don't think we should consider a job with only failed pod as available but yes for succeeded.

Reply:
Job is a particularly tricky API type to handle. It really depends on the completion pattern I guess; if it's indexed, having one succeeded pod does not say much, and for those with workqueues, we might never see it's completion.

Reply:
Taking a peek at the job spec and decides in accordance with the job controller behavior seems like too much work though 😣